commit 1b70cdddcdb9ef32ae1014098351964ee6b62cda (refs/changes/56/1856/4)
Author: Jason King <jason.king@joyent.com>
Date:   2017-05-01T19:42:19-05:00 (2 years, 5 months ago)
    
    NAPI-375 Multiple filter use on IP search endpoint seems to only heed first filter

diff --git a/lib/endpoints/search.js b/lib/endpoints/search.js
index 5dde889..bd3fe33 100644
--- a/lib/endpoints/search.js
+++ b/lib/endpoints/search.js
@@ -60,6 +60,16 @@ function searchIPsFilterArgs(params) {
             continue;
         }
 
+        /*
+         * We skip "ip" since we only ever fetch matching IPs from Moray.
+         * Additionally, beyond being superfluous, the saved IP in the stream
+         * (sipn_ip) is an object, while the passed in value will be a string,
+         * so the comparison will fail.
+         */
+        if (key === 'ip') {
+            continue;
+        }
+
         if (SIP_LIST_ARGS[key] === true) {
             list_args[key] = params[key];
         } else {
@@ -102,11 +112,13 @@ SearchIPNetworkStream.prototype._transform = function (net, _enc, done) {
      * Determine whether or not we should even consider this network.
      */
     if (self.sipn_ipkind !== net.subnetStart.kind()) {
-        return done();
+        done();
+        return;
     }
 
     if (!net.subnet.contains(self.sipn_ip)) {
-        return done();
+        done();
+        return;
     }
 
     /*
@@ -128,7 +140,8 @@ SearchIPNetworkStream.prototype._transform = function (net, _enc, done) {
     mod_ip.get(getOpts, function (err, ipobj) {
         var sip, prop;
         if (err || !ipobj) {
-            return done(err);
+            done(err);
+            return;
         }
 
         sip = ipobj.serialize();
@@ -138,16 +151,18 @@ SearchIPNetworkStream.prototype._transform = function (net, _enc, done) {
             }
 
             if (!sip.hasOwnProperty(prop)) {
-                return done();
+                done();
+                return;
             }
 
-            if (self.sipn_args.prop !== sip.prop) {
-                return done();
+            if (self.sipn_args[prop] !== sip[prop]) {
+                done();
+                return;
             }
         }
 
         self.push(sip);
-        return done();
+        done();
     });
 };
 
diff --git a/lib/util/constants.js b/lib/util/constants.js
index 4c00b14..ff26ead 100644
--- a/lib/util/constants.js
+++ b/lib/util/constants.js
@@ -53,7 +53,7 @@ var MESSAGES = {
     PROV_END_TYPE_MISMATCH: 'provision_end_ip and subnet must both be ' +
         'IPv4 or IPv6 addresses',
 
-    SEARCH_NO_NETS: 'No networks found containing that IP address',
+    SEARCH_NO_NETS: 'No networks found matching search criteria',
     STR: 'must be a string',
     UNKNOWN_PARAMS: 'Unknown parameters',
     VLAN_USED: 'VLAN ID is already in use',
diff --git a/test/lib/ip.js b/test/lib/ip.js
index 5329038..ab5f477 100644
--- a/test/lib/ip.js
+++ b/test/lib/ip.js
@@ -101,6 +101,21 @@ function listIPs(t, opts, callback) {
         common.afterAPIlist.bind(null, t, opts, callback));
 }
 
+function searchIPs(t, opts, callback) {
+    var client = opts.client || mod_client.get();
+
+    assert.object(t, 't');
+    assert.string(opts.ip, 'opts.ip');
+    assert.object(opts.params, 'opts.params');
+
+    opts.type = TYPE;
+    opts.id = 'ip';
+
+    log.debug({ params: opts.params }, 'search IPs');
+
+    client.searchIPs(opts.ip, opts.params,
+        common.afterAPIlist.bind(null, t, opts, callback));
+}
 
 /**
  * Update an IP and compare the output
@@ -131,5 +146,6 @@ module.exports = {
     get: getIP,
     freeIP: freeIPrecord,
     list: listIPs,
+    search: searchIPs,
     update: updateIP
 };
diff --git a/test/unit/search-ips.test.js b/test/unit/search-ips.test.js
index b39414a..c5374d5 100644
--- a/test/unit/search-ips.test.js
+++ b/test/unit/search-ips.test.js
@@ -18,7 +18,9 @@ var constants = require('../../lib/util/constants');
 var fmt = require('util').format;
 var h = require('./helpers');
 var mod_err = require('../../lib/util/errors');
+var mod_ip = require('../lib/ip');
 var mod_net = require('../lib/net');
+var mod_nic = require('../lib/nic');
 var mod_nicTag = require('../lib/nic-tag');
 var mod_server = require('../lib/server');
 var mod_uuid = require('node-uuid');
@@ -230,7 +232,65 @@ test('IP not in any networks', function (t) {
     });
 });
 
+test('Search', function (t) {
+    var params = {
+        belongs_to_type: 'zone',
+        belongs_to_uuid: mod_uuid.v4(),
+        owner_uuid: mod_uuid.v4()
+    };
+
+    var ipObj = {};
+
+    h.copyParams(params, ipObj);
+
+    ipObj.reserved = false;
+    ipObj.free = false;
+
+    t.test('create ip', function (t2) {
+        mod_nic.provision(t2, {
+            net: NETS[2].uuid,
+            params: params,
+            partialExp: params,
+            fillInMissing: true
+        }, function (_, res) {
+            ipObj.network_uuid = res.network_uuid;
+            ipObj.ip = res.ip;
+            t2.end();
+        });
+    });
+
+    t.test('find - just ip', function (t2) {
+        mod_ip.search(t2, {
+            ip: ipObj.ip,
+            params: {},
+            present: [ ipObj ]
+        });
+    });
 
+    t.test('find - ip & belongs_to_uuid', function (t2) {
+        mod_ip.search(t2, {
+            ip: ipObj.ip,
+            params: {
+                belongs_to_uuid: ipObj.belongs_to_uuid
+            },
+            present: [ ipObj ]
+        });
+    });
+
+    t.test('find - ip & belongs_to_uuid doesn\'t match', function (t2) {
+        mod_ip.search(t2, {
+            ip: ipObj.ip,
+            params: {
+                belongs_to_uuid: mod_uuid.v4()
+            },
+            expCode: 404,
+            expErr: {
+                code: 'ResourceNotFound',
+                message: constants.msg.SEARCH_NO_NETS
+            }
+        });
+    });
+});
 
 // --- Teardown
 
